### ISSUE TITLE
Inference with `instanceof` and generic unions is unsound

### ISSUE THREAD

**Issue by wchargin** [facebook/flow#6741]
Inference with `instanceof` and generic unions is unsound

If `x: T | MyClass<T>`, then Flow infers from `x instanceof MyClass`
that `x: MyClass<T>`. This is unsound: `T` could be `MyClass<U>` for
some other `U`, so that in fact `x: MyClass<MyClass<U>>`.

[Here is an example of how this can go wrong:][1]

```js
// @flow
class Box<T> {
  +field: T;
  constructor(x) {
    this.field = x;
  }
}

function asBox<T>(x: T | Box<T>): Box<T> {
  if (x instanceof Box) {
    // Here, `x` is refined to be `Box<T>`. This is unsound: `T` could
    // well be `Box<U>` for some other `U`. Example below.
    return x;
  } else {
    // whatever
    return new Box(x);
  }
}

const stringStringBox: Box<Box<string>> = asBox(new Box("wat"));  // unsound!
stringStringBox.field.field.substr(0);  // runtime error
```

[1]: https://flow.org/try/#0PTAEAEDMBsHsHcBQBjaBDAzh0AhWAPAHgBUA+UAb0VFAGpIBLAU2gBMAuUYgbmtGVgA7DABcATgFdkI2GIAU+AJSU+NEQAsGGAHSMWrUAF5Q+XjQC+iS4kgTB0hkNCY8RMgs7FQAH1wESpIqcrgEqNAyQoAqgDMIiaPZMsJGuylQ0NCCgABJMYkwANKAABvjFMdj5jIJMBjKgAEZMJSFkxdpcmthaoHYYsHYcJcTlAhJsqqBZ8CzQjc3FIQCqpOWQsqD9ALbNsBp5JUvtoACi+GhbAA7QzU1w8NqT+SISYoImZqDmoCwYzekZabqNAiJgANzyTyYLzeoBq8D8+AUik+lmsAjim3EsQA5gBlbGCHGuYL+EKiMS40jkYwuAhyeGIuQAIngIOZihRmTAfQGglYAEJEBTcQTKUTXLpmGwpfptBgJA0KXIAAxcqZgSSCEQMHY-MRiWSIIA

**Comment by jbrown215** [comment]
cc @samwgoldman

**Comment by jbrown215** [comment]
After speaking to @samwgoldman , it's possible that we are refining x to `Box<any>`. I'll take a look at this!

**Comment by samwgoldman** [comment]
Yeah, see https://github.com/facebook/flow/blob/7057f045695764d512a483ba2d5102bdaa06390a/src/typing/flow_js.ml#L3995

Worth experimenting with using `instantiate_poly` instead, which creates an implicit type argument (i.e., some fresh tvar instead of `any`). But I haven't thought about this very hard and I wonder about perf.

**Comment by wchargin** [comment]
> Worth experimenting with using instantiate_poly instead, which creates an implicit type argument (i.e., some fresh tvar instead of any).

Makes sense; sounds like right semantics. 

Would be nice for `string | Box<T>` to refine to `Box<T>` instead of `Box<?>` under the appropriate refinement. Not clear to me whether `inatantiate_poly` would admit that? Would prefer that incompleteness over this unsoundness, though.

(Apologies for brevity; on mobile.)
